pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,222 @@
+// cross-resource attribute references. See
+// https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-cross-resource-referencing.md
+// for more information
+type AttributeReferencer interface {
+	// Build retrieves and builds the referenced attribute
+	Build(getter LocalObjectGetter) (value string, err error)
+
+	// Assign accepts a resource object, and assigns the given value to the corresponding property
+	Assign(mgd Managed, value string) error
+}
+
+// A ManagedReferenceResolver resolves the references to other managed resources
+type ManagedReferenceResolver interface {
+	ResolveReferences(context.Context, Managed) error
+}
+
+// ReferenceResolver resolves implements ManagedReferenceResolver interface
+type ReferenceResolver struct {
+	reader client.Reader
+}
+
+// NewReferenceResolver returns a new ReferenceResolver
+func NewReferenceResolver(c client.Reader) *ReferenceResolver {
+	return &ReferenceResolver{c}
+}
+
+// ResolveReferences resolves references made to other managed resources
+func (r *ReferenceResolver) ResolveReferences(ctx context.Context, mgd Managed) (err error) {
+	// this recovers from potential panics during execution of
+	// AttributeReferencer methods
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf("paniced while resolving references: %v", r)
+		}
+	}()
+
+	referencers, err := findRefResolverFields(mgd, false)
+	if err != nil {
+		return err
+	}
+
+	getter := r.getterFunc(ctx, mgd)
+	err = validateReferencesAreReady(getter, referencers)
+	if err != nil {
+		return err
+	}
+
+	for _, referencer := range referencers {
+		val, err := referencer.Build(getter)
+		if err != nil {
+			return err
+		}
+
+		err = referencer.Assign(mgd, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getterFunc returns a function to retrieve local references
+func (r *ReferenceResolver) getterFunc(ctx context.Context, mgd Managed) LocalObjectGetter {
+	return func(ref corev1.LocalObjectReference, obj runtime.Object) error {
+		err := r.reader.Get(ctx,
+			types.NamespacedName{
+				Namespace: mgd.GetNamespace(),
+				Name:      ref.Name,
+			},
+			obj)
+
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+// findRefResolverFields recursively finds all non-nil fields in a struct and its sub-structs
+// that implement AttributeReferencer and have `managed_resource:"reference_resolver"` tag
+func findRefResolverFields(obj interface{}, objHasTheRightTag bool) ([]AttributeReferencer, error) { // nolint:gocyclo
+
+	v := reflect.ValueOf(obj)
+	result := []AttributeReferencer{}
+
+	// ignore nil fields
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return nil, nil
+	}
+
+	if objHasTheRightTag {
+		if ar, implements := obj.(AttributeReferencer); implements {
+			return []AttributeReferencer{ar}, nil
+		} else if v.Kind() == reflect.Struct {
+			return nil, errors.Errorf("the field has the 'reference_resolver' tag, but has not implemented AttributeReferencer interface")
```

Comment:
This seems like a condition that we'd ideally test for when implementing the type, rather than at runtime. I typically feel like it's appropriate to panic rather than return an error in these cases where a binary is fundamentally broken due to programmer error.

Reply:
My understanding is that the `reference_resolver` is intended to increase readability of our API types, rather than being necessary to implement this functionality. If that is the case, and if this reference resolution code would be simpler without the tag (relying only on checking whether types satisfied the `AttributeReferencer` interface), I think I'd prefer we omit the tag.

Reply:
The tag doesn't adding any significant complexity, other than just figuring that a field is indeed intended to be a `AttributeReferencer`. Beside increasing readability it is used to manifest the intention of the developer.

Without this tag, it is impossible for the `ReferenceResolver` to decide whether  it should error when a field has some but not all of `AttributeReferencer` methods. Similarly